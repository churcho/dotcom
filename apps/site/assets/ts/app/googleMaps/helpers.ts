// @ts-ignore
import stopMarker from "../../../static/images/icon-map-stop-marker.svg";
// @ts-ignore
import stationMarker from "../../../static/images/icon-map-station-marker.svg";
// @ts-ignore
import selectedStopMarker from "../../../static/images/icon-map-stop-marker-hover.svg";
// @ts-ignore
import selectedStationMarker from "../../../static/images/icon-map-station-marker-hover.svg";
// @ts-ignore
import currentLocationMarker from "../../../static/images/icon-current-location-marker.svg";
import { iconSize } from "../../../js/google-map/helpers";
import mapStyles from "../../../js/google-map/styles";
import { MarkerData, MapData } from "./__googleMaps";

export const stopIcon = (selected: boolean): string =>
  selected ? selectedStopMarker : stopMarker;
export const stationIcon = (selected: boolean): string =>
  selected ? selectedStationMarker : stationMarker;

export const buildIconFromSVG = (
  svg: string,
  size: string
): google.maps.Icon => {
  const sizeInt = iconSize(size);

  const encoded = window.btoa(svg);

  return {
    url: `data:image/svg+xml;base64, ${encoded}`,
    scaledSize: new window.google.maps.Size(sizeInt, sizeInt),
    origin: new window.google.maps.Point(0, 0),
    labelOrigin: new window.google.maps.Point(sizeInt / 2, sizeInt / 2)
  };
};

export const buildMarkerIcon = (
  data: MarkerData,
  selected: boolean
): google.maps.Icon | undefined => {
  if (data.icon) {
    if (data.icon!.includes("station")) {
      return buildIconFromSVG(stationIcon(selected), data.size);
    }
    if (data.icon!.includes("stop")) {
      return buildIconFromSVG(stopIcon(selected), data.size);
    }
    if (data.icon!.includes("current-location")) {
      return buildIconFromSVG(currentLocationMarker, data.size);
    }
  }
  return undefined;
};

export const setMapDefaults = (
  map: google.maps.Map,
  initialData: MapData
): void => {
  const options = {
    center: new window.google.maps.LatLng(
      initialData.default_center.latitude,
      initialData.default_center.longitude
    ),
    ...initialData.dynamic_options
  };
  const styles = mapStyles as google.maps.MapTypeStyle[];
  // Hide stop icons on the transit layer
  styles.push({
    featureType: "transit",
    elementType: "labels.icon",
    stylers: [{ visibility: "off" }]
  });
  map.setOptions({ styles, ...options });
  map.setZoom(initialData.zoom || 17);

  if (initialData.layers && initialData.layers.transit) {
    const layer = new window.google.maps.TransitLayer();
    layer!.setMap(map);
  }
};
