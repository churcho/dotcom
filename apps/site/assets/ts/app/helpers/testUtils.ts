// eslint-disable-next-line import/prefer-default-export
export const createReactRoot = (): void => {
  document.body.innerHTML =
    '<div><div id="react-root"><div id="test"></div></div></div>';
};
