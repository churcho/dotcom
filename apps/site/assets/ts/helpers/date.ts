// Returns a date in the form of "August 7, 2019"
export const formattedDate = (unformatted: string): string => {
  const [year, month, day] = unformatted
    .split(/-/)
    .map(part => Number.parseInt(part, 10));

  // Remember that months in JS are 0-indexed for some reason, hence "month - 1".
  const parsedDate = new Date(year, month - 1, day);
  return parsedDate.toLocaleDateString("en-US", {
    year: "numeric",
    month: "long",
    day: "numeric",
    timeZone: "UTC"
  });
};

// Returns a time in the form of "05:16 AM"
export const shortTime = (date: string): string =>
  new Date(date).toLocaleTimeString("en-US", {
    hour: "2-digit",
    minute: "2-digit"
  });

export default formattedDate;
