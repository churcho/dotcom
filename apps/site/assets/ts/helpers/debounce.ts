// Debounce is meant to be used with event listeners. Use it to delay a call
// until after a burst of events, such as keyboard input.
//
//   usage:
//
//     const callback = (ev: Event) => {}
//     document.addEventListener("keyup", debounce(callback, 250))
//

const debounce = (callback: Function, delay: number): EventListener => {
  let timer: number | undefined;

  return (...args) => {
    window.clearTimeout(timer);
    timer = window.setTimeout(() => callback(...args), delay);
  };
};

export default debounce;
