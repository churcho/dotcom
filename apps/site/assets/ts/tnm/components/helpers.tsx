import React, { KeyboardEvent } from "react";

export const handleEnterKeyPress = (
  e: KeyboardEvent,
  onClick: Function
): void => {
  if (e.key === "Enter") {
    onClick();
  }
};

export const renderSvg = (className: string, svgText: string): JSX.Element => (
  <span
    className={className}
    // eslint-disable-next-line react/no-danger
    dangerouslySetInnerHTML={{ __html: svgText }}
  />
);
