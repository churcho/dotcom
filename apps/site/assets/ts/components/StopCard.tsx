import React, { ReactElement, KeyboardEvent } from "react";
import { Direction as DirectionType, Route, Stop } from "../__v3api";
import { clickStopCardAction, Dispatch } from "../tnm/state";
import Direction from "./Direction";
import renderSvg from "../helpers/render-svg";
import { accessibleIcon } from "../helpers/icon";
// @ts-ignore
import stationSymbol from "../../static/images/icon-circle-t-small.svg";

interface Props {
  stop: Stop;
  directions: DirectionType[];
  route: Route;
  dispatch: Dispatch;
}

const renderStopIcon = (stop: Stop): JSX.Element =>
  stop["station?"] ? (
    renderSvg("m-tnm-sidebar__stop-marker", stationSymbol)
  ) : (
    <></>
  );

const handleKeyPress = (
  e: KeyboardEvent<HTMLDivElement>,
  onClick: Function
): void => {
  if (e.key === "Enter") {
    onClick();
  }
};

export const StopCard = ({
  stop,
  directions,
  route,
  dispatch
}: Props): ReactElement<HTMLElement> | null => {
  const onClick = (): void => dispatch(clickStopCardAction(stop.id));

  const key = `${route.id}-${stop.id}`;

  return (
    <div
      role="button"
      tabIndex={0}
      className="m-tnm-sidebar__route-stop"
      onClick={onClick}
      onKeyPress={e => handleKeyPress(e, onClick)}
    >
      <div className="m-tnm-sidebar__stop-info">
        <a href={stop.href} className="m-tnm-sidebar__stop-name">
          {renderStopIcon(stop)}
          {stop.name}
          {!!stop.accessibility.length &&
            !stop.accessibility.includes("unknown") &&
            accessibleIcon("m-tnm-sidebar__stop-accessible")}
        </a>
        <div className="m-tnm-sidebar__stop-distance">{stop.distance}</div>
      </div>
      {directions.map(direction => (
        <Direction
          key={`${key}-${direction.direction_id}`}
          direction={direction}
          route={route}
          stopId={stop.id}
        />
      ))}
    </div>
  );
};

export default StopCard;
