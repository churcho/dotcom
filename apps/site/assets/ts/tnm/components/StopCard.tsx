import React, { ReactElement, KeyboardEvent } from "react";
import { Stop, Route } from "./__tnm";
import { clickStopCardAction, Dispatch } from "../state";
import { Direction, directionIsEmpty } from "./Direction";
import { renderSvg } from "./helpers";
// @ts-ignore
import stationSymbol from "../../../static/images/icon-circle-t-small.svg";
// @ts-ignore
import accessibleIcon from "../../../static/images/icon-accessible-default.svg";

interface Props {
  stop: Stop;
  route: Route;
  dispatch: Dispatch;
}

export const stopIsEmpty = (stop: Stop): boolean =>
  stop.directions.every(directionIsEmpty);

const renderStopIcon = (stop: Stop): JSX.Element =>
  stop["station?"] ? (
    renderSvg("m-tnm-sidebar__stop-marker", stationSymbol)
  ) : (
    <></>
  );

const handleKeyPress = (
  e: KeyboardEvent<HTMLDivElement>,
  onClick: Function
): void => {
  if (e.key === "Enter") {
    onClick();
  }
};

export const StopCard = ({
  stop,
  route,
  dispatch
}: Props): ReactElement<HTMLElement> | null => {
  const key = `${route.id}-${stop.id}`;

  if (stopIsEmpty(stop)) {
    return null;
  }

  const onClick = (): void => dispatch(clickStopCardAction(stop.id));

  return (
    <div
      role="button"
      tabIndex={0}
      className="m-tnm-sidebar__route-stop"
      onClick={onClick}
      onKeyPress={e => handleKeyPress(e, onClick)}
    >
      <div className="m-tnm-sidebar__stop-info">
        <a href={stop.href} className="m-tnm-sidebar__stop-name">
          {renderStopIcon(stop)}
          {stop.name}
          {!!stop.accessibility.length &&
            !stop.accessibility.includes("unknown") &&
            renderSvg("m-tnm-sidebar__stop-accessible", accessibleIcon)}
        </a>
        <div className="m-tnm-sidebar__stop-distance">{stop.distance}</div>
      </div>
      {stop.directions.map(direction => (
        <Direction
          key={`${key}-${direction.direction_id}`}
          direction={direction}
          route={route}
          stopId={stop.id}
        />
      ))}
    </div>
  );
};

export default StopCard;
