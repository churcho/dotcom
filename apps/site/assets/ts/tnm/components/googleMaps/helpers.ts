// @ts-ignore
import stopMarker from "../../../../static/images/icon-map-stop-marker.svg";
// @ts-ignore
import stationMarker from "../../../../static/images/icon-map-station-marker.svg";
// @ts-ignore
import selectedStopMarker from "../../../../static/images/icon-map-stop-marker-hover.svg";
// @ts-ignore
import selectedStationMarker from "../../../../static/images/icon-map-station-marker-hover.svg";
// @ts-ignore
import currentLocationMarker from "../../../../static/images/icon-current-location-marker.svg";
import { iconSize } from "../../../../js/google-map/helpers";

import { MarkerData } from "../__tnm";

interface GoogleMapMarkerSVG {
  url: string;
  scaledSize: google.maps.Size;
  origin: google.maps.Point;
  anchor: google.maps.Point;
  labelOrigin: google.maps.Point;
}

export const stopIcon = (selected: boolean): string =>
  selected ? selectedStopMarker : stopMarker;
export const stationIcon = (selected: boolean): string =>
  selected ? selectedStationMarker : stationMarker;

export const buildIconFromSVG = (
  svg: string,
  size: string
): GoogleMapMarkerSVG => {
  const sizeInt = iconSize(size);

  const encoded = window.btoa(svg);

  return {
    url: `data:image/svg+xml;base64, ${encoded}`,
    scaledSize: new window.google.maps.Size(sizeInt, sizeInt),
    origin: new window.google.maps.Point(0, 0),
    anchor: new window.google.maps.Point(sizeInt / 2, sizeInt / 2),
    labelOrigin: new window.google.maps.Point(sizeInt / 2, sizeInt / 2)
  };
};

export const buildMarkerIcon = (
  data: MarkerData,
  selected: boolean
): GoogleMapMarkerSVG | undefined => {
  if (data.icon) {
    if (data.icon!.includes("station")) {
      return buildIconFromSVG(stationIcon(selected), data.size);
    }
    if (data.icon!.includes("stop")) {
      return buildIconFromSVG(stopIcon(selected), data.size);
    }
    if (data.icon!.includes("current-location")) {
      return buildIconFromSVG(currentLocationMarker, data.size);
    }
  }
  return undefined;
};
