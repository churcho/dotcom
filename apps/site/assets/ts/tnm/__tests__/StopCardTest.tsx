import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import StopCard from "../components/StopCard";
import { createReactRoot } from "../../app/helpers/testUtils";
import {
  TNMDirection,
  TNMHeadsign,
  TNMRoute,
  TNMStop,
  TNMTime
} from "../components/__tnm";

/* eslint-disable typescript/camelcase */

const time: TNMTime = {
  scheduled_time: ["4:30", " ", "PM"],
  prediction: null,
  delay: 0
};

const headsign: TNMHeadsign = {
  name: "Headsign",
  times: [time],
  train_number: null
};

const direction: TNMDirection = {
  direction_id: 0,
  headsigns: [headsign]
};

const stop: TNMStop = {
  accessibility: ["wheelchair"],
  address: "123 Main St., Boston MA",
  bike_storage: [],
  closed_stop_info: null,
  fare_facilities: [],
  "has_charlie_card_vendor?": false,
  "has_fare_machine?": false,
  id: "stop-id",
  "is_child?": false,
  latitude: 41.0,
  longitude: -71.0,
  name: "Stop Name",
  note: null,
  parking_lots: [],
  "station?": true,
  distance: "238 ft",
  directions: [direction],
  href: "/stops/stop-id"
};

const route: TNMRoute = {
  alert_count: 0,
  direction_destinations: ["Outbound Destination", "Inbound Destination"],
  direction_names: ["Outbound", "Inbound"],
  id: "route-id",
  name: "Route Name",
  header: "Route Header",
  long_name: "Route Long Name",
  description: "Route Description",
  stops: [stop],
  type: 1
};

it("it renders a stop card", () => {
  createReactRoot();
  const tree = renderer
    .create(<StopCard stop={stop} route={route} dispatch={() => {}} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("returns null if stop has no schedules", () => {
  const stopWithoutSchedules: TNMStop = {
    ...stop,
    directions: [{ ...direction, headsigns: [] }]
  };
  const tree = renderer
    .create(
      <StopCard stop={stopWithoutSchedules} route={route} dispatch={() => {}} />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("it selects stop when the card is clicked", () => {
  createReactRoot();

  const mockDispatch = jest.fn();

  const wrapper = shallow(
    <StopCard stop={stop} route={route} dispatch={mockDispatch} />
  );

  wrapper.find(".m-tnm-sidebar__route-stop").simulate("click");
  expect(mockDispatch).toHaveBeenCalledWith({
    type: "CLICK_STOP_CARD",
    payload: { stopId: "stop-id" }
  });
});

it("it selects stop when the card is selected via keyboard", () => {
  createReactRoot();

  const mockDispatch = jest.fn();

  const wrapper = shallow(
    <StopCard stop={stop} route={route} dispatch={mockDispatch} />
  );

  wrapper
    .find(".m-tnm-sidebar__route-stop")
    .simulate("keyPress", { key: "Enter" });
  expect(mockDispatch).toHaveBeenCalledWith({
    type: "CLICK_STOP_CARD",
    payload: { stopId: "stop-id" }
  });
});

it("it does nothing when the keyboard event is not enter", () => {
  createReactRoot();

  const mockDispatch = jest.fn();

  const wrapper = shallow(
    <StopCard stop={stop} route={route} dispatch={mockDispatch} />
  );

  wrapper
    .find(".m-tnm-sidebar__route-stop")
    .simulate("keyPress", { key: "Tab" });
  expect(mockDispatch).not.toHaveBeenCalled();
});
