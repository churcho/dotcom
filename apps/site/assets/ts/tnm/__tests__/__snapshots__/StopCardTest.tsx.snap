// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a stop card 1`] = `
<div
  className="m-tnm-sidebar__stop"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="m-tnm-sidebar__stop-info"
  >
    <a
      className="m-tnm-sidebar__stop-name"
      href="/stops/9983"
    >
      Stuart St @ Charles St S
      <span
        className="m-tnm-sidebar__stop-accessible"
        dangerouslySetInnerHTML={
          Object {
            "__html": "SVG",
          }
        }
      />
    </a>
    <div
      className="m-tnm-sidebar__stop-distance"
    >
      238 ft
    </div>
  </div>
  <div>
    <div
      className="m-tnm-sidebar__direction"
    >
      <div
        className="m-tnm-sidebar__direction-name u-small-caps"
      >
        Outbound
      </div>
    </div>
    <div
      className="m-tnm-sidebar__headsign-schedule m-tnm-sidebar__headsign-schedule--condensed"
    >
      <div
        className="m-tnm-sidebar__headsign"
      >
        <div
          className="m-tnm-sidebar__headsign-name m-tnm-sidebar__headsign-name--large"
        >
          Watertown
        </div>
        <div
          className="m-tnm-sidebar__via"
        >
          via Copley (Express)
        </div>
      </div>
      <div
        className="m-tnm-sidebar__schedules"
      >
        <div
          className="m-tnm-sidebar__schedule"
        >
          <div
            className="m-tnm-sidebar__time"
          >
            <div
              className="m-tnm-sidebar__time-number"
            >
              14
            </div>
            <div
              className="m-tnm-sidebar__time-mins"
            >
              min
            </div>
          </div>
        </div>
        <div
          className="m-tnm-sidebar__schedule"
        >
          <div
            className="m-tnm-sidebar__time"
          >
            <div
              className="m-tnm-sidebar__time-number"
            >
              44
            </div>
            <div
              className="m-tnm-sidebar__time-mins"
            >
              min
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`returns null if stop has no schedules 1`] = `null`;
